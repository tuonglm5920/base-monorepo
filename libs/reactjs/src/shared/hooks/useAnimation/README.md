# Overview

The `useAnimation` hook is a custom React hook designed to facilitate the use of the Web Animations API. It provides a straightforward way to define and control animations within React components. By taking an `AnimationDefinition`, this hook returns an `AnimationController` object, offering various methods to manage the animation lifecycle.

# Options

### Parameters

##### Generic Type

- **Args**: `AnyRecord | void` (default: `void`) - A type for arguments passed to the animation's keyframe function, allowing for dynamic keyframe generation.

##### Hook Parameters

- **props**: `AnimationDefinition<Args>` - The definition of the animation, including keyframes and options.

### Return Value

- `AnimationController<Args>`: An instance of AnimationController, providing methods and properties to control the animation.

# Examples

### Simple

```jsx
const App = () => {
  // 1. Define your animation in WAAPI way
  const animate = useAnimation({
    keyframe: [{ transform: "rotate(0deg)" }, { transform: "rotate(720deg)" }],
    options: {
      duration: 1000,
      easing: "ease-in-out",
    },
  });

  return (
    <button
      // 2. You have to pass animate to element you want to control
      ref={animate}
      onClick={() => {
        // 3. And play it!
        animate.play();
      }}
    >
      Click Me!
    </button>
  );
};
```

### Dynamic keyframe

```jsx
const Index: FC = () => {
  // Define argument type
  const animate = useAnimation<{ x: number; y: number }>({
    keyframe: (prev, args) => [
      // You can get current style from 1st argument
      { transform: prev.transform },
      // Get passed position from 2nd argument
      { transform: `translate(${args.x}px, ${args.y}px)` },
    ],
    options: {
      duration: 400,
      easing: 'ease-in-out',
    },
  });

  useEffect(() => {
    // If you click somewhere, the circle follows you!

    const onClick = (event: MouseEvent) => {
      // Pass mouse position when animate
      animate.play({
        args: { x: event.clientX, y: event.clientY },
      });
    };
    window.addEventListener('click', onClick);
    return () => {
      window.removeEventListener('click', onClick);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <div
      ref={animate}
      style={{
        position: 'fixed',
        border: 'solid 0.1rem #135569',
        borderRadius: '50%',
        height: '6rem',
        width: '6rem',
        top: '-3rem',
        left: '-3rem',
      }}
    />
  );
};

```

### Progress tracking

````jsx
const App = () => {
  const [value, setValue] = useState(0);
  const animate = useAnimationFunction<number>(
    ({ progress }, arg) => {
      // Do anything here!
      setValue(progress * arg);
    },
    {
      duration: 600,
      easing: "ease-in-out",
    }
  );
  useEffect(() => {
    animate.play({ args: 100 });
  }, []);

  return <progress value={value} max={100} style={{ width: 600 }} />;
};
```

# Warning

### <u>Partial keyframes are not supported</u> error was thrown

```typescript
// valid
const animate = useAnimation([{ transform: "translate3d(0px, 0, 0)" }, { transform: "translate3d(400px, 0, 0)" }], { duration: 800, easing: "ease-in-out" });
// invalid
const animate = useAnimation({ transform: "translate3d(400px, 0, 0)" }, { duration: 800, easing: "ease-in-out" });

// valid
const animate = useAnimation(
  [
    { transform: "translateX(0px)", fill: "blue" },
    { transform: "translateX(100px)", fill: "red" },
    { transform: "translateX(0px)", fill: "blue" },
  ],
  { duration: 800, easing: "ease-in-out" },
);
// invalid
const animate = useAnimation([{ transform: "translateX(0px)" }, { transform: "translateX(100px)", fill: "red" }, { fill: "blue" }], { duration: 800, easing: "ease-in-out" });
````
