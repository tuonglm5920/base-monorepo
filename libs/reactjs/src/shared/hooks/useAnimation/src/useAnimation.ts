import { AnyRecord } from 'typescript-utilities';
import { WebAnimationApi } from 'utilities';
import { useLatestPropsRef } from '../../useLatestPropsRef';
import { useMount } from '../../useMount';
import { useOnce } from '../../useOnce';
import { AnimationController, BaseAnimationController } from './types/AnimationController';
import { AnimationDefinition } from './types/AnimationDefinition';
import { normalizeKeyframe } from './utils/normalizeKeyframe';
import {
  ConfigOfWebAnimationApiInstance,
  WebAnimationApiInstancesManager,
} from './utils/WebAnimationApiInstancesManager';

const webAnimationApiInstancesManager = new WebAnimationApiInstancesManager();

/**
 * A basic hook to use the Web Animations API within a React component. It simplifies the process of creating
 * and controlling animations based on the provided animation definition.
 *
 * @template Args - A type that extends AnyRecord or void. This type is used for the arguments passed to the animation's keyframe function,
 *                   allowing for dynamic generation of keyframes based on the provided arguments.
 *
 * @param {AnimationDefinition<Args>} props - The definition of the animation, including keyframes and options.
 * @returns {AnimationController<Args>} An instance of AnimationController, providing methods and properties to control the animation.
 *
 * @see {@link AnimationController} for details on the returned controller object.
 */
export const useAnimation = <Args extends AnyRecord | void = void>(
  props: AnimationDefinition<Args>,
): AnimationController<Args> => {
  const argsRef = useLatestPropsRef(props);

  const [handle, mount] = useOnce((): [AnimationController<Args>, () => () => void] => {
    let $el: Element | null = null;
    let configOfWebAnimationApiInstance: ConfigOfWebAnimationApiInstance | undefined;

    const createWebAnimationApiInstance = (args: Args): WebAnimationApi | undefined => {
      if (!$el) {
        return;
      }
      const { keyframe = [], options = {} } = argsRef.current ?? [];

      const prevActive = configOfWebAnimationApiInstance;
      configOfWebAnimationApiInstance = {
        _keyframes: normalizeKeyframe($el, keyframe, args),
        _options: options,
      };

      return webAnimationApiInstancesManager.createInstance($el, configOfWebAnimationApiInstance, prevActive);
    };
    const cleanWebAnimationInstance = (): void => {
      if (configOfWebAnimationApiInstance) {
        webAnimationApiInstancesManager.deleteInstanceByConfig(configOfWebAnimationApiInstance);
      }
    };

    const controller: AnimationController<Args> = Object.assign(
      (ref: Element | null) => {
        $el = ref;
        if (!$el) {
          cleanWebAnimationInstance();
        }
      },
      {
        reverse: () => webAnimationApiInstancesManager.getInstanceByConfig(configOfWebAnimationApiInstance)?.reverse(),
        cancel: () => webAnimationApiInstancesManager.getInstanceByConfig(configOfWebAnimationApiInstance)?.cancel(),
        finish: () => webAnimationApiInstancesManager.getInstanceByConfig(configOfWebAnimationApiInstance)?.finish(),
        pause: () => webAnimationApiInstancesManager.getInstanceByConfig(configOfWebAnimationApiInstance)?.pause(),
        setPlaybackRate: rate => {
          webAnimationApiInstancesManager.getInstanceByConfig(configOfWebAnimationApiInstance)?.setRate(rate);
        },
        waitFor: event => {
          webAnimationApiInstancesManager.getInstanceByConfig(configOfWebAnimationApiInstance)?.waitFor(event);
        },
        play: options => {
          const webAnimationApiInstance = createWebAnimationApiInstance('args' in options ? options.args : {});
          webAnimationApiInstance?.play({});
        },
        setTime: time => {
          let webAnimationApiInstance =
            configOfWebAnimationApiInstance &&
            webAnimationApiInstancesManager.getInstanceByConfig(configOfWebAnimationApiInstance);
          if (!webAnimationApiInstance) {
            const { keyframe } = argsRef.current;
            if (typeof keyframe === 'function') {
              return;
            }
            // Init animation in setTime to start animation without calling play
            webAnimationApiInstance = createWebAnimationApiInstance(undefined!);
          }
          webAnimationApiInstance?.setTime(time);
        },
      } as BaseAnimationController<Args>,
    );

    return [controller, (): (() => void) => cleanWebAnimationInstance];
  });

  // eslint-disable-next-line react-hooks/exhaustive-deps
  useMount(mount);

  return handle;
};
