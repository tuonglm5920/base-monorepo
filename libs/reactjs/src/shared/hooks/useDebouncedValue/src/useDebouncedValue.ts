import { useEffect, useRef, useState } from 'react';

interface DebouncedOptions {
  /** The amount of time, in milliseconds, to delay the debounced value. */
  timeoutMs?: number;
}

/**
 * Provides a debounced value of the given input. The debounced value will only reflect the latest value
 * if no changes have been made within the specified timeout duration.
 *
 * @template T The type of the value being debounced.
 * @param {T} value The value to debounce.
 * @param {DebouncedOptions} options Configuration options for the debounce behavior.
 * @param {number} [options.timeoutMs=500] The amount of time, in milliseconds, to delay the debounced value.
 *                                          Defaults to 500ms if not specified.
 * @returns {T} The debounced value, which updates only after the specified timeout has elapsed since the
 *              last change to the input value.
 *
 * @example
 * const [value, setValue] = useState('');
 * const debouncedValue = useDebouncedValue(value, { timeoutMs: 300 });
 *
 * // `debouncedValue` will reflect the latest `value` only if it hasn't changed for at least 300ms.
 */
export const useDebouncedValue = <T>(value: T, { timeoutMs = 500 }: DebouncedOptions): T => {
  const [state, setState] = useState<T>(value);
  const stateRef = useRef<T>(state);

  useEffect(() => {
    if (stateRef.current === value) {
      return;
    }

    const timeout = setTimeout(() => {
      stateRef.current = value;
      setState(value);
    }, timeoutMs);

    return (): void => {
      clearTimeout(timeout);
    };
  }, [value, timeoutMs]);

  return state;
};
