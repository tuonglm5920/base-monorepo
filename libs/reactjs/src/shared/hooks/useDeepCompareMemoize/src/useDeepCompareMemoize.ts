import { dequal } from 'dequal';
import { DependencyList, useRef } from 'react';

/**
 * Custom hook that performs a deep comparison on the provided dependencies and returns a memoized version
 * of the dependencies.
 *
 * @param {DependencyList} value An array of dependencies to be deeply compared and memoized.
 *
 * @returns {DependencyList} A memoized dependency list that only changes if a deep comparison indicates changes.
 */
export const useDeepCompareMemoize = (value: DependencyList): DependencyList => {
  const ref = useRef<DependencyList>([]);

  if (!dequal(value, ref.current)) {
    ref.current = value;
  }

  return ref.current;
};
