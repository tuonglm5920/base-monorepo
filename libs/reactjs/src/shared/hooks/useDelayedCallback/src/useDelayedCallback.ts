import { useRef } from 'react';
import { useMount } from '../../useMount';

/**
 * Provides a delayed callback function that can be invoked after a specified delay.
 *
 * @param {Function} callback - The callback function to be executed after the delay.
 * @param {number} delay - The time to wait before the callback is executed, in milliseconds.
 * @returns {() => void} A function that, when called, will set up the delayed execution of the callback.
 *                        Calling this function multiple times will reset the timer.
 *
 * @example
 * const handleDelayedLog = useDelayedCallback(() => console.log('Delayed Log'), 1000);
 *
 * // When `handleDelayedLog` is called, it will log 'Delayed Log' to the console after 1 second.
 * handleDelayedLog();
 */
export const useDelayedCallback = (callback: Function, delay: number): (() => void) => {
  const timeoutRef = useRef<NodeJS.Timeout | undefined>();

  useMount(() => {
    return (): void => {
      if (timeoutRef.current != null) {
        clearTimeout(timeoutRef.current);
      }
    };
  });

  const callbackWithDelay = (): void => {
    const timeoutId = setTimeout(() => {
      callback();
      timeoutRef.current = undefined;
    }, delay);

    timeoutRef.current = timeoutId;
  };

  return callbackWithDelay;
};
