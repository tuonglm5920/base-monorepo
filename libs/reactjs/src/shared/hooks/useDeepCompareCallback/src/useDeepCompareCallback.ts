import { DependencyList, useCallback } from 'react';
import { useDeepCompareMemoize } from '../../useDeepCompareMemoize';

/**
 * `useDeepCompareCallback` is a custom hook that memoizes a callback function with deep comparison of the dependencies.
 * Usage note: only use this if `deps` are objects or arrays that contain objects. Otherwise you should just use React.useMemo.
 * @template T Callback function type.
 * @param {T} callback - The callback function to be memoized.
 * @param {React.DependencyList} dependencies - An array of dependencies for which the memoization will be deep compared.
 * @returns {T} - Returns the memoized callback function.
 *
 * @example ```typescript
  const memoizedCallback = useDeepCompareCallback(() => {
    console.log('Callback is called');
  }, [complexDependency]);
  ```
 */
export const useDeepCompareCallback = <T extends (...args: any[]) => any>(
  callback: T,
  dependencies: DependencyList,
): ReturnType<typeof useCallback> => {
  // eslint-disable-next-line react-hooks/exhaustive-deps
  return useCallback(callback, useDeepCompareMemoize(dependencies));
};
