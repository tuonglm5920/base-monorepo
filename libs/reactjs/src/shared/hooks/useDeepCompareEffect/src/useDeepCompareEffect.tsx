import { DependencyList, EffectCallback, useEffect } from 'react';
import { useDeepCompareMemoize } from '../../useDeepCompareMemoize';

/**
 * `useDeepCompareEffect` is a custom hook similar to `useEffect` but with deep comparison
 * of the dependencies list.
 * Usage note: only use this if `deps` are objects or arrays that contain objects. Otherwise you should just use React.useMemo.
 *
 * @param {React.EffectCallback} effect - A function that contains imperative, possibly
 *    effectful code. This function will be deep compared against changes in the dependencies list.
 * @param {React.DependencyList} dependencies - An array of dependencies for which the effect
 *    will be executed when there are deep changes.
 * @returns {void}
 *
 * @example
 * useDeepCompareEffect(() => {
 *   // Some effectful logic here...
 * }, [complexDependency]);
 */
export const useDeepCompareEffect = (effect: EffectCallback, dependencies: DependencyList): void => {
  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(effect, useDeepCompareMemoize(dependencies));
};
