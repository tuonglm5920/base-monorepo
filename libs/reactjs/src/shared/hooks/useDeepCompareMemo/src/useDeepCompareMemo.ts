import { useMemo } from 'react';
import { useDeepCompareMemoize } from '../../useDeepCompareMemoize';

/**
 * A custom React hook that memoizes the result of a factory function based on deep comparison of its dependencies.
 * This hook is similar to `React.useMemo`, but it uses deep comparison instead of reference equality to determine if dependencies have changed.
 * Usage note: only use this if `deps` are objects or arrays that contain objects. Otherwise you should just use React.useMemo.
 *
 * @template T - The type of the value returned by the factory function.
 * @param {() => T} factory - A function that returns a value to be memoized.
 * @param {React.DependencyList} dependencies - An array of dependencies to determine when the memoized value should be recomputed.
 * @returns {T} - Returns the memoized value.
 *
 * @example ```typescript
  const expensiveValue = useDeepCompareMemo(() => computeExpensiveValue(a, b), [a, b]);
  ```
 */
export const useDeepCompareMemo = <T>(factory: () => T, dependencies: React.DependencyList): T => {
  // eslint-disable-next-line react-hooks/exhaustive-deps
  return useMemo(factory, useDeepCompareMemoize(dependencies));
};
