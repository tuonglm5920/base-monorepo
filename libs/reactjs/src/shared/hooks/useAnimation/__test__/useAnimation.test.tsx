/* eslint-disable @nx/workspace/no-strings-outside-text-component */
import { expect } from '@jest/globals';
import { fireEvent, render } from '@testing-library/react';
import { FC, useEffect } from 'react';
import { renderToStaticMarkup, renderToString } from 'react-dom/server';
import { useAnimation } from '../src/useAnimation';

describe('useAnimation', () => {
  const Comp: FC = () => {
    const animate = useAnimation({
      keyframe: { width: 100 },
      options: { duration: 600, easing: 'ease-in-out' },
    });
    useEffect(() => {
      animate.play({});
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    return (
      <div ref={animate} style={{ width: 50 }}>
        test
      </div>
    );
  };

  it('should succeed with renderToString', () => {
    expect(renderToString(<Comp />)).toMatchSnapshot();
  });

  it('should succeed with renderToStaticMarkup', () => {
    expect(renderToStaticMarkup(<Comp />)).toMatchSnapshot();
  });
});

describe('CircleComponent with useAnimation hook', () => {
  const CircleComponent: FC = () => {
    const animate = useAnimation<{ x: number; y: number }>({
      keyframe: (prev, args) => [{ transform: prev.transform }, { transform: `translate(${args.x}px, ${args.y}px)` }],
      options: {
        duration: 400,
        easing: 'ease-in-out',
      },
    });

    useEffect(() => {
      const onClick = (event: MouseEvent): void => {
        animate.play({ args: { x: event.clientX, y: event.clientY } });
      };
      window.addEventListener('click', onClick);
      return () => {
        window.removeEventListener('click', onClick);
      };
    }, [animate]);

    return (
      <div
        ref={animate}
        style={{
          position: 'fixed',
          border: 'solid 0.1rem #135569',
          borderRadius: '50%',
          height: '6rem',
          width: '6rem',
          top: '-3rem',
          left: '-3rem',
        }}
      />
    );
  };
  const mockTime = 500;
  const mockDuration = 1000;
  const mockAnimations = new Map();

  class MockAnimation {
    public effect: MockKeyframeEffect;
    public options: number | KeyframeAnimationOptions | undefined;
    public playState: AnimationPlayState;
    public playbackRate: number;
    public eventTarget: EventTarget;
    constructor(effect: MockKeyframeEffect, options: number | KeyframeAnimationOptions | undefined) {
      this.effect = effect;
      this.options = options;
      this.playState = 'idle';
      this.playbackRate = Number.MAX_SAFE_INTEGER;
      this.eventTarget = new EventTarget();
    }

    public play = (): void => {
      this.playState = 'running';
    };

    public pause = (): void => {
      this.playState = 'paused';
    };

    public reverse = (): void => {
      this.playbackRate = -this.playbackRate;
    };

    public cancel = (): void => {
      this.playState = 'idle';
    };

    public updatePlaybackRate(newRate: number): void {
      this.playbackRate = newRate;
    }

    public finish = (): void => {
      this.playState = 'finished';
      // Simulate the end state of the animation
      if (this.effect.keyframes && Array.isArray(this.effect.keyframes) && this.effect.keyframes.length > 0) {
        const lastKeyframe = this.effect.keyframes[this.effect.keyframes.length - 1];
        if (typeof lastKeyframe['transform'] === 'string') {
          this.effect.element.style.transform = lastKeyframe['transform'];
        }
      }
      this.dispatchEvent(new Event('finish'));
    };

    // Delegate event listener methods to the internal EventTarget instance
    public addEventListener(
      type: string,
      listener: EventListenerOrEventListenerObject,
      options?: boolean | AddEventListenerOptions,
    ): void {
      this.eventTarget.addEventListener(type, listener, options);
    }

    public removeEventListener(
      type: string,
      listener: EventListenerOrEventListenerObject,
      options?: boolean | EventListenerOptions,
    ): void {
      this.eventTarget.removeEventListener(type, listener, options);
    }

    public dispatchEvent(event: Event): boolean {
      return this.eventTarget.dispatchEvent(event);
    }
  }

  class MockKeyframeEffect {
    public element: HTMLElement;
    public keyframes: PropertyIndexedKeyframes | Keyframe[] | null;
    public options: number | KeyframeAnimationOptions | undefined;

    public getComputedTiming(): ComputedEffectTiming {
      return {
        endTime: mockDuration,
        activeDuration: mockDuration,
        localTime: mockTime,
        progress: 0.5,
      };
    }
    constructor(
      element: HTMLElement,
      keyframes: Keyframe[] | PropertyIndexedKeyframes | null,
      options: number | KeyframeAnimationOptions | undefined,
    ) {
      this.element = element;
      this.keyframes = keyframes;
      this.options = options;
    }
  }

  HTMLElement.prototype.animate = function (
    keyframes: Keyframe[] | PropertyIndexedKeyframes | null,
    options?: number | KeyframeAnimationOptions,
  ): any {
    const effect = new MockKeyframeEffect(this, keyframes, options);
    const animation = new MockAnimation(effect, options);
    mockAnimations.set(this, animation); // Store the animation instance
    return animation;
  };

  it('moves to click position when clicked', async () => {
    const { container } = render(<CircleComponent />);
    const circle = container.firstChild as HTMLElement;

    // Simulate a click
    fireEvent.click(document, { clientX: 100, clientY: 150 });

    // Manually call finish on the mock animation
    const animation = mockAnimations.get(circle);
    animation.finish();

    // Assert the final state
    expect(circle.style.transform).toBe('translate(100px, 150px)');
  });
});
