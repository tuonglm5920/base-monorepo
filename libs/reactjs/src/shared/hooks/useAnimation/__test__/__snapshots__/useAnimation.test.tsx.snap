// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useAnimation should succeed with renderToStaticMarkup 1`] = `"<div style="width:50px">test</div>"`;

exports[`useAnimation should succeed with renderToString 1`] = `"<div style="width:50px">test</div>"`;
